### PR TITLE
change(limit-count): ensure redis cluster name is set correctly

### PR DESCRIPTION
**Pull request by han6565**
…ess test prompting ' ttl: ERR invalid passwor'

### What this PR does / why we need it:

during the stress test, I found that if the limit-count plugin configures multiple different redis-cluster data sources, the following error will appear failed to get redis `plugin-limit-countxx.xx.xx.xxroute122` ttl: ERR invalid password so I Fixed it

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [ ] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by tokers** [comment]
Could you please add some test cases to cover the scene that you said?

File: apisix/plugins/limit-count/limit-count-redis-cluster.lua [hunk 1]
```diff
@@ -32,7 +32,7 @@ local mt = {
 
 local function new_redis_cluster(conf)
     local config = {
-        name = "apisix-redis-cluster",
+        name = "apisix-redis-cluster-" .. conf.name,
```
Comment (spacewander) [inline apisix/plugins/limit-count/limit-count-redis-cluster.lua:35]: I am confused by the solution. What is the relation between the password and name?

**Comment by han6565** [comment]
> Could you please add some test cases to cover the scene that you said?

ok,I can paste some of my test configuration and test results 
in etcd one route like this:
{
    "id": "347382218030582023",
    "create_time": 1616585537,
    "update_time": 1616590678,
    "uris": [
        "/test"
    ],
    "name": "limit",
    "methods": [
        "GET",
        "HEAD",
        "POST",
        "PUT",
        "DELETE",
        "OPTIONS",
        "PATCH"
    ],
    "plugins": {
        "limit-count": {
            "count": 1000,
            "time_window": 1,
            "rejected_code": 503,
            "policy": "redis-cluster",
            "redis_cluster_nodes": [
                "cluster1-node1:6379",
                "cluster1-node2:6379",
                "cluster1-node3:6379"
            ],
            "redis_password": "cluster1-password"
        }
    },
    "upstream": {
        "nodes": [
            {
                "host": "myip",
                "port": 9999,
                "weight": 1
            }
        ],
        "timeout": {
            "connect": 6000,
            "read": 6000,
            "send": 6000
        },
        "type": "roundrobin",
        "pass_host": "pass"
    },
    "status": 1
}
other route like this:
{
    "id": "347382218030582024",
    "create_time": 1616585538,
    "update_time": 1616590679,
    "uris": [
        "/test2"
    ],
    "name": "limit2",
    "methods": [
        "GET",
        "HEAD",
        "POST",
        "PUT",
        "DELETE",
        "OPTIONS",
        "PATCH"
    ],
    "plugins": {
        "limit-count": {
            "count": 1000,
            "time_window": 1,
            "rejected_code": 503,
            "policy": "redis-cluster",
            "redis_cluster_nodes": [
                "cluster2-node1:6379",
                "cluster2-node2:6379",
                "cluster2-node3:6379"
            ],
            "redis_password": "cluster2-password"
        }
    },
    "upstream": {
        "nodes": [
            {
                "host": "myip",
                "port": 9999,
                "weight": 1
            }
        ],
        "timeout": {
            "connect": 6000,
            "read": 6000,
            "send": 6000
        },
        "type": "roundrobin",
        "pass_host": "pass"
    },
    "status": 1
}

then I stress test two routes at the same time，i use the go-stress-testing  test the url like this:
./go-stress-testing-mac -c 20 -n 100000 -u http://apisix_ip:9080/test
./go-stress-testing-mac -c 20 -n 100000 -u http://apisix_ip:9080/test2 

you can see my error.log it output errror  like this:
2021/03/26 11:07:03 [error] 28109#28109: *6443 [lua] limit-count.lua:174: phase_func(): failed to limit req: failed to get redis `plugin-limit-count10.73.42.1route122` ttl: ERR invalid password, client: 10.73.42.1, server: , request: "GET /test HTTP/1.1", host: "172.20.4.213:9080"
but the password is correct。
I will upload the error log
[error.log](https://github.com/apache/apisix/files/6209169/error.log)

File: apisix/plugins/limit-count/limit-count-redis-cluster.lua (hunk 1 above)
Comment (han6565) [reply apisix/plugins/limit-count/limit-count-redis-cluster.lua:35]: one name Mapping one routes，if all redis_cluster config the same name. When concurrent lrucache can not find the correct redis_cluster in pool. 
This is my personal opinion ，maybe it is wrong.😄

Comment (spacewander) [reply apisix/plugins/limit-count/limit-count-redis-cluster.lua:35]: I read through the `resty.rediscluster`. Look like this library uses a module-level `slot_cache`, and the key is the `config.name`.

So it is no doubted that different configurations with the same name will collide.

What about using `"apisix-redis-cluster-" .. tostring(conf)` directly?
Better to add a comment about this change as future maintainer may not read through the library.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/plugins/limit-count/limit-count-redis-cluster.lua [hunk 2]
```diff
@@ -32,7 +32,8 @@ local mt = {
 
 local function new_redis_cluster(conf)
     local config = {
-        name = "apisix-redis-cluster-" .. conf.name,
+	-- can set different name for different redis cluster
```
Comment (spacewander) [inline apisix/plugins/limit-count/limit-count-redis-cluster.lua:35]: style: indent

File: apisix/plugins/limit-count.lua [hunk 3]
```diff
@@ -103,6 +103,9 @@ local schema = {
                         redis_timeout = {
                             type = "integer", minimum = 1, default = 1000,
                         },
+                        redis_cluster_name = {
```
Comment (spacewander) [inline apisix/plugins/limit-count.lua:106]: We should update the doc too.

File: apisix/plugins/limit-count.lua [hunk 4]
```diff
@@ -103,6 +103,9 @@ local schema = {
...
+                            type = "string", default = "apisix-redis-cluster",
+                        },
                     },
                     required = {"redis_cluster_nodes"},
```
Comment (spacewander) [inline apisix/plugins/limit-count.lua:110]: For the reason I have said, the `redis_cluster_name` should be required.

**Review by spacewander — CHANGES_REQUESTED** [review]

File: apisix/plugins/limit-count.lua (hunk 4 above)
Comment (spacewander) [reply apisix/plugins/limit-count.lua:110]: the redis_cluster_name should be required

File: docs/en/latest/plugins/limit-count.md [hunk 5]
```diff
@@ -48,6 +48,7 @@ Limit request rate by a fixed number of requests in a given time window.
...
 | redis_cluster_nodes | array   | optional             |               |                                                                                                         | When using `redis-cluster` policy，This property is a list of addresses of Redis cluster service nodes.                                                                                                                                                                                                    |
+| redis_cluster_name  | string  | optional             |               |                                                                                                         | When using `redis-cluster` policy，This property is the name of Redis cluster service nodes.                                                                                                                                                                                                                   |
```
Comment (spacewander) [inline docs/en/latest/plugins/limit-count.md:51]: Should update doc and test too

File: apisix/plugins/limit-count.lua [hunk 6]
```diff
@@ -103,8 +103,11 @@ local schema = {
                         redis_timeout = {
                             type = "integer", minimum = 1, default = 1000,
                         },
+                        redis_cluster_name = {
+                            type = "string",
+                        },
                     },
-                    required = {"redis_cluster_nodes"},
+                    required = {"redis_cluster_nodes","redis_cluster_name"},
```
Comment (Firstsawyou) [inline apisix/plugins/limit-count.lua:110]: Bad style, it should be like this:
```suggestion
                    required = {"redis_cluster_nodes", "redis_cluster_name"},
```

File: docs/en/latest/plugins/limit-count.md (hunk 5 above)
Comment (Firstsawyou) [inline docs/en/latest/plugins/limit-count.md:51]: It should be a comma in English, and the sentence does not end. `This` should be lowercase.
```suggestion
| redis_cluster_name  | string  | optional             |               |                                                                                                         | When using `redis-cluster` policy, this property is the name of Redis cluster service nodes.                                                                                                                                                                                                                   |
```

**Review by Firstsawyou — CHANGES_REQUESTED** [review]

**Comment by spacewander** [comment]
Don't forget to update the example: https://github.com/apache/apisix/blob/master/docs/en/latest/plugins/limit-count.md

```
"plugins": {
        "limit-count": {
            "count": 2,
            "time_window": 60,
            "rejected_code": 503,
            "key": "remote_addr",
            "policy": "redis-cluster",
            "redis_cluster_nodes": [
              "127.0.0.1:5000",
              "127.0.0.1:5001"
            ]
        }
    },
```

File: docs/en/latest/plugins/limit-count.md [hunk 7]
```diff
@@ -48,6 +48,7 @@ Limit request rate by a fixed number of requests in a given time window.
...
 | redis_cluster_nodes | array   | optional             |               |                                                                                                         | When using `redis-cluster` policy，This property is a list of addresses of Redis cluster service nodes.                                                                                                                                                                                                    |
+| redis_cluster_name  | string  | optional             |               |                                                                                                         | When using `redis-cluster` policy, this property is the name of Redis cluster service nodes.                                                                                                                                                                                                                   |
```
Comment (spacewander) [inline docs/en/latest/plugins/limit-count.md:51]: Need to change the "optional".

File: docs/zh/latest/plugins/limit-count.md [hunk 8]
```diff
@@ -51,6 +51,7 @@ title: limit-count
...
 | redis_cluster_nodes | array   | 可选         |               |                                                                                                         | 当使用 `redis-cluster` 限速策略时，该属性是 Redis 集群服务节点的地址列表。                                                                                                                                                                                                                                                                                                                                            |
+| redis_cluster_name  | string  | 可选         |               |                                                                                                         | 当使用 `redis-cluster` 限速策略时，该属性是 Redis 集群服务节点的名称。                                                                                                                                                                                                                                                                                                                                            |
```
Comment (spacewander) [inline docs/zh/latest/plugins/limit-count.md:54]: Ditto

File: docs/en/latest/plugins/limit-count.md (hunk 7 above)
Comment (spacewander) [reply docs/en/latest/plugins/limit-count.md:51]: @han6565
This need to be addressed

**Review by spacewander — APPROVED** [review]

**Review by Firstsawyou — APPROVED** [review]

File: docs/en/latest/plugins/limit-count.md [hunk 9]
```diff
@@ -35,20 +35,20 @@ Limit request rate by a fixed number of requests in a given time window.
...
+| redis_cluster_nodes | array   | optional                     |               |                                                                                                         | When using `redis-cluster` policy，This property is a list of addresses of Redis cluster service nodes.                                                                                                                                                                                                    |
+| redis_cluster_name  | string  | required for `redis-cluster` |               |                                                                                                         | When using `redis-cluster` policy, this property is the name of Redis cluster service nodes.                                                                                                                                                                                                                   |
```
Comment (tokers) [inline docs/en/latest/plugins/limit-count.md:51]: What about "required when `policy` is `redis-cluster`".

File: docs/zh/latest/plugins/limit-count.md [hunk 10]
```diff
@@ -38,20 +38,20 @@ title: limit-count
...
+| redis_cluster_nodes | array   | 可选                 |               |                                                                                                         | 当使用 `redis-cluster` 限速策略时，该属性是 Redis 集群服务节点的地址列表。                                                                                                                                                                                                                                                                                                                                            |
+| redis_cluster_name  | string  | `redis-cluster` 必须 |               |                                                                                                         | 当使用 `redis-cluster` 限速策略时，该属性是 Redis 集群服务节点的名称。                                                                                                                                                                                                                                                                                                                                            |
```
Comment (tokers) [inline docs/zh/latest/plugins/limit-count.md:54]: What about "当 `policy` 为 `redis-cluster` 时必填"

**Review by tokers — APPROVED** [review]

**Pull request merged**